legacy/messaging/src/main/java/org/jboss/as/messaging/MigrateOperation.java

```diff
@@ -0,0 +1,370 @@
+    }
+
+    @Override
+    public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
+        if (!describe && context.getRunningMode() != RunningMode.ADMIN_ONLY) {
+            throw MessagingLogger.ROOT_LOGGER.migrateOperationAllowedOnlyInAdminOnly();
+        }
+
+        boolean addLegacyEntries = ADD_LEGACY_ENTRIES.resolveModelAttribute(context, operation).asBoolean();
+        boolean dryRun = DRY_RUN.resolveModelAttribute(context, operation).asBoolean();
+
+        // node containing the description (list of add operations) of the legacy subsystem
+        final ModelNode legacyModelAddOps = new ModelNode();
+        // preserve the order of insertion of the add operations for the new subsystem.
+        final Map<PathAddress, ModelNode> migrationOperations = new LinkedHashMap<PathAddress, ModelNode>();
+
+        // invoke an OSH to describe the legacy messaging subsystem
+        describeLegacyMessagingResources(context, legacyModelAddOps);
+        // invoke an OSH to add the messaging-activemq extension
+        // FIXME: this does not work it the extension :add is added to the migrationOperations directly (https://issues.jboss.org/browse/WFCORE-323)
+        addMessagingActiveMQExtension(context, migrationOperations, describe, dryRun);
+        context.addStep(new OperationStepHandler() {
+            @Override
+            public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
+                // transform the legacy add operations and put them in migrationOperations
+                transformResources(legacyModelAddOps, migrationOperations, addLegacyEntries);
+                // put the /subsystem=messaging:remove operation
+                removeMessagingSubsystem(migrationOperations);
+
+                if (describe) {
+                    // :describe-migration operation
+
+                    // for describe-migration operation, do nothing and return the list of operations that would
+                    // be executed in the composite operation
+                    context.getResult().set(migrationOperations.values());
+                } else {
+                    // :migrate operation
+
+                    if (!dryRun) {
+                        // invoke an OSH on a composite operation with all the migration operations
+                        migrateSubsystems(context, migrationOperations);
+                    }
+                    context.getResult().set(!dryRun);
+                }
+            }
+        }, MODEL);
+    }
+
+
+    /**
+     * It's possible that the extension is already present. In that case, this method does nothing.
+     */
+    private void addMessagingActiveMQExtension(OperationContext context, Map<PathAddress, ModelNode> migrationOperations, boolean describe, boolean dryRun) {
+        Resource root = context.readResourceFromRoot(PathAddress.EMPTY_ADDRESS);
```

Comment:
This can be context.readResourceFromRoot(PathAddress.EMPTY_ADDRESS, false) which will save cloning the entire model. Even with the 'false' param your logic at L198 will work properly, as discussed in the method javadoc:

```
     * Use the {@code recursive} parameter to avoid the expense of making read-only copies of large portions of the
     * resource tree. If {@code recursive} is {@code false}, the returned resource will only have placeholder resources
     * for immediate children. Those placeholder resources will return an empty
     * {@link org.jboss.as.controller.registry.Resource#getModel() model} and will not themselves have any children.
     * Their presence, however, allows the caller to see what immediate children exist under the target resource.
```